### PR TITLE
Close #85 - Feature/mysql protocol argument tcp socket pipe memory

### LINKED ISSUE #85: No way to connect via TCP?
##### SUMMARY

I receive an error message when using the `mysql_db` module to connect to a MySQL Docker container.

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME

mysql_db

##### ANSIBLE VERSION

```paste below
ansible 2.10.3
  config file = /home/sturm/devops/ansible.cfg
  configured module search path = ['/home/sturm/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /home/sturm/.local/lib/python3.8/site-packages/ansible
  executable location = /home/sturm/.local/bin/ansible
  python version = 3.8.5 (default, Jul 28 2020, 12:59:40) [GCC 9.3.0]
```

##### CONFIGURATION

```paste below
DEFAULT_HOST_LIST(/home/sturm/devops/ansible.cfg) = ['/home/sturm/devops/inventories/main']
DEFAULT_LOAD_CALLBACK_PLUGINS(/home/sturm/devops/ansible.cfg) = True
DEFAULT_ROLES_PATH(/home/sturm/devops/ansible.cfg) = ['/home/sturm/devops/roles']
DEFAULT_STDOUT_CALLBACK(/home/sturm/devops/ansible.cfg) = yaml
DEFAULT_VAULT_PASSWORD_FILE(/home/sturm/devops/ansible.cfg) = /home/sturm/devops/.vault_pass
HOST_KEY_CHECKING(/home/sturm/devops/ansible.cfg) = False
INTERPRETER_PYTHON(/home/sturm/devops/ansible.cfg) = /usr/bin/python3
```

##### OS / ENVIRONMENT

Host:
Windows 10 v2004 with WSL2 running Ubuntu 20.04
Target:
Ubuntu 18.04

##### STEPS TO REPRODUCE

1. Ensure that a MySQL Docker container is up and running on the target with some data.
2. Write a simple playbook using the task below. (Replace the variables with your own info.)
3. Run the playbook and watch the error message appear:
```
fatal: [skyubuntu]: FAILED! => changed=false
  msg: |-
    mysqldump: [Warning] Using a password on the command line interface can be insecure.
    mysqldump: Got error: 2002: Can't connect to local MySQL server through socket '/var/run/mysqld/mysqld.sock' (2) when trying to connect
```

Here's the task for calling the mysqldump process:
```yaml
  tasks:
    - name: Dump Sky Schedule database
      community.mysql.mysql_db:
        name: "{{ db_database }}"
        state: dump
        target: /tmp/{{ db_database }}.sql
        login_host: localhost
        login_port: 3306
        login_user: "{{ vault_db_username }}"
        login_password: "{{ vault_db_password }}"
```

##### EXPECTED RESULTS

I expect the module to connect to the MySQL server via TCP.

##### ACTUAL RESULTS

The module appears to be trying to connect to the MySQL container via a unix socket. Per [this SO answer](https://stackoverflow.com/a/34489879/2379279), the only way to connect to a MySQL container is to specifically set the protocol to TCP. I see no way to do so with `mysql_db`.

```paste below
 ___________________________________
< TASK [Dump Sky Schedule database] >
 -----------------------------------
        \   ^__^
         \  (oo)\_______
            (__)\       )\/\
                ||----w |
                ||     ||

[WARNING]: Module did not set no_log for unsafe_login_password
fatal: [skyubuntu]: FAILED! => changed=false
  msg: |-
    mysqldump: [Warning] Using a password on the command line interface can be insecure.
    mysqldump: Got error: 2002: Can't connect to local MySQL server through socket '/var/run/mysqld/mysqld.sock' (2) when trying to connect
```

### PR DESCRIPTION
**Pull request by BuhtigithuB**
##### SUMMARY

Support for mysql --protocol=<PROTOCOL>

Close #85

##### ISSUE TYPE

- Feature Pull Request

### CODE REVIEW

File: plugins/modules/mysql_query.py [hunk 1]
```diff
@@ -43,7 +43,8 @@
     - Where passed queries run in a single transaction (C(yes)) or commit them one-by-one (C(no)).
     type: bool
     default: no
-seealso:
+
+see also:
```
Comment (Andersson007) [inline plugins/modules/mysql_query.py:47]:
```suggestion
see also:
```
this is not a typo and this key is used by parsers to generate the documentation, so please revert the change:)

File: plugins/modules/mysql_db.py [hunk 2]
```diff
@@ -316,7 +335,10 @@
 
 from ansible.module_utils.basic import AnsibleModule
 from ansible_collections.community.mysql.plugins.module_utils.database import mysql_quote_identifier
-from ansible_collections.community.mysql.plugins.module_utils.mysql import mysql_connect, mysql_driver, mysql_driver_fail_msg, mysql_common_argument_spec
+from ansible_collections.community.mysql.plugins.module_utils.mysql import (
```
Comment (Andersson007) [inline plugins/modules/mysql_db.py:338]:
```suggestion
from ansible_collections.community.mysql.plugins.module_utils.mysql import (
```
please don't mix several unrelated changes in one PR as it makes things harder to review. Could you please eliminate all unrelated changes (like format fixes)?

**Review by Andersson007 — COMMENTED** [review]
@BuhtigithuB thanks for the PR!

1. Use https://github.com/ansible-collections/community.mysql/blob/main/plugins/doc_fragments/mysql.py instead of copying the option description to every module, if the new argument is applicable for all the modules
2. Use https://github.com/ansible-collections/community.mysql/blob/main/plugins/module_utils/mysql.py if the new argument is applicable for all the modules
3. Add a changelog fragment https://docs.ansible.com/ansible/latest/community/development_process.html#creating-a-changelog-fragment
4. Move all unrelated changes (like formatting ones) to a separate PR
5. Fix CI (you can see details in the test box at the bottom, click on `Details` where there are red crosses). You can also run tests locally before pushing https://github.com/ansible/community-docs/blob/main/integration_tests_quick_start_guide.rst
6. Add integration tests https://github.com/ansible/community-docs/blob/main/integration_tests_quick_start_guide.rst

If you have any questions, feel free to ask.

Thank you!

File: plugins/modules/mysql_query.py (hunk 1 above)
Comment (BuhtigithuB) [reply plugins/modules/mysql_query.py:47]: I was asking myself about that one

File: plugins/modules/mysql_db.py (hunk 2 above)
Comment (BuhtigithuB) [reply plugins/modules/mysql_db.py:338]: yes

**Comment by BuhtigithuB** [comment]
:( 

I am afraid it won't work as there is no protocol parameters in pymysql : 

https://github.com/PyMySQL/PyMySQL/blob/main/pymysql/connections.py#L167

I was also scratching my head about that one...

Such a pita to improve on this one...

**Comment by Andersson007** [comment]
> :(
> 
> I am afraid it won't work as there is no protocol parameters in pymysql :
> 
> https://github.com/PyMySQL/PyMySQL/blob/main/pymysql/connections.py#L167
> 
> I was also scratching my head about that one...
> 
> Such a pita to improve on this one...

yeah... we should think of another possible solution, we could continue that in the issue.
Anyway, you could convert this PR to a formatting fix one. It would be much appreciated. What do you think? (changelog fragment is not needed in that case)

**Comment by Andersson007** [comment]
@BuhtigithuB would you like to complete this PR by making it fixing only the formatting issues?

**Comment by Andersson007** [comment]
@BuhtigithuB feel free to create a PR to fix the formatting issues you found along the way, thanks!
